File: Extra Credit/EEAAO/Graph.java
import java.util.ArrayList;

public class Graph {
    private ArrayList<ArrayList<Integer>> adjList;
    private int V;
    
    // Initialize a graph with V vertices
    Graph(int V) {
        this.V = V;
        adjList = new ArrayList<>(V);
        for (int i = 0; i < V; i++) {
            adjList.add(new ArrayList<>());
        }
    }

    // Add an edge from src to dest
    void addEdge(int src, int dest) {
        adjList.get(src).add(dest);
    }

    int[][] adjacencyListToMatrix() {
        int[][] matrix = new int[V][V];
        for (int i = 0; i < V; i++) {
            for (int dest : adjList.get(i)) {
                matrix[i][dest] = 1;
            }
        }
        return matrix;
    }

    ArrayList<ArrayList<Integer>> adjacencyMatrixToList(int[][] matrix) {
        ArrayList<ArrayList<Integer>> adjList = new ArrayList<>();
        for (int i = 0; i < V; i++) {
            ArrayList<Integer> list = new ArrayList<>();
            for (int j = 0; j < V; j++) {
                if (matrix[i][j] == 1) list.add(j);
            }
            adjList.add(list);
        }
        return adjList;
    }

    Graph reverseEdges() {
        Graph reverseGraph = new Graph(V);
        for (int i = 0; i < V; i++) {
            for (int dest : adjList.get(i)) {
                reverseGraph.addEdge(dest, i);
            }
        }
        return reverseGraph;
    }
}

File: Extra Credit/EEAAO/Graph.py
class Graph:
    def __init__(self, V):
        self.V = V
        self.adjList = [[] for _ in range(V)]

    def addEdge(self, src, dest):
        self.adjList[src].append(dest)

    def adjacencyListToMatrix(self):
        matrix = [[0 for _ in range(self.V)] for _ in range(self.V)]
        for i in range(self.V):
            for dest in self.adjList[i]:
                matrix[i][dest] = 1
        return matrix

    def adjacencyMatrixToList(self, matrix):
        adjList = [[] for _ in range(self.V)]
        for i in range(self.V):
            for j in range(self.V):
                if matrix[i][j] == 1:
                    adjList[i].append(j)
        return adjList

    def reverseEdges(self):
        reverseGraph = Graph(self.V)
        for i in range(self.V):
            for dest in self.adjList[i]:
                reverseGraph.addEdge(dest, i)
        return reverseGraph.adjList